### DISCUSSION TITLE
Problem with sys.stdout.reconfigure and TextIOWrapper partially unknown type

### DISCUSSION THREAD

**Discussion by DervishD** [microsoft/pyright#9645] [Q&A] [answered]
Hi there!

I have the following code in a program, which I think it's pretty standard:
```
import sys
from io import TextIOWrapper

if sys.stdout and isinstance(sys.stdout, TextIOWrapper):
    sys.stdout.reconfigure(encoding='utf-8')
```

I'm getting  an error from `PyRight` (well, `Pylance`, used under `VSCode`), to wit: Type of "stdout" is partially unknown.  Type of "stdout" is "TextIOWrapper[Unknown]"). This was not happening the last time I worked in this code, a few months ago, so I don't know if it is a bug or not.

Thing is, I don't know how to get rid of this error except using `cast(TextIOWrapper, sys.stdout)` and I cannot do this because in fact `sys.stdout` can be redirected and not be of class `TextIOWrapper`. Also, I don't entirely understand why the cast works here suppressing the error, and I prefer not to use a solution I don't understand.

Am I doing anything wrong here? How can I suppress the error without resorting to dubious code constructs? Is it a bug in `Pyright` or `typeshed` or anything like it?

I've googled about this problem for days without finding anything and help from Copilot and ChatGPT has been quite useless.

Thanks in advance for any information about this. I can't believe I'm the only one wanting to force `UTF-8` on output **AND** using `Pyright` 😆

**Comment by erictraut** [comment, marked as answer]
Pyright is working correctly here. I've confirmed that the behavior hasn't changed in pyright for a long time. I suspect that the type definitions for `stdio` have changed over time, which would explain why you observed a difference in behavior. These type definitions are provided in typeshed and bundled with pyright and pylance.

The type of `stdout` is defined as `TextIO | MaybeNone`. The `TextIOWrapper` type is a subclass of `TextIO`, but it is defined as generic. It takes a single type argument. When you use the `isinstance` check to narrow the type of `TextIO`, pyright has no way of inferring the type argument for `TextIOWrapper`, so it infers `Unknown`. If you know the intended type of the type argument, you could use a `cast` within the conditional block. Unfortunately, `isinstance` doesn't allow you to pass a specialized type directly as its second argument.

It appears that you're using strict-mode type checking. By default pyright does not report unknown or partially-unknown types. If you don't want to see these errors reported, you can switch from "strict" to "standard" type checking mode.

**Comment by DervishD** [comment]
@erictraut thanks A LOT for the explanation. I was focused unto TextIOWrapper and forgot about TextIO.
Yes, I'm using "strict" mode, but I cannot switch to "standard", for now it's mandatory to use "strict" mode. Looks like I'm not doing anything wrong so I may just silent the warning, since the code is OK. I may replace the `isinstance()` by a `try` block and ignore `AttributeError` if `reconfigure()` is not defined for the actual class `sys.stdio` belongs to at runtime, or any other solution.

And yes, probably the problem is in typeshed, apparently it has been updated since I last worked on the code.

Thanks a lot, really, I was worried about some unintended side-effect in the code I was using, even though as I wrote it seems pretty standard.
